File: py/Sqrt_x.py
class Solution:
    def mySqrt(self, x: int) -> int:
        if x == 0:
            return 0
        if x < 4:
            return 1
        res = 2 * self.mySqrt(x / 4)
        if (res + 1) * (res + 1) <= x and (res + 1) * (res + 1) >= 0:
            return res + 1
        return res

if __name__ == '__main__':
    s = Solution()
    print(s.mySqrt(8))

File: c_plus/Sqrt_x.cpp
#include <iostream>
using namespace std;

class Solution {
public:
    int mySqrt(int x) {
        if (x == 0) return 0;
        if (x < 4) return 1;
        long res = 2 * mySqrt(x / 4);
        if ((res + 1) * (res + 1) <= x) return res + 1;
        return res;
    }
};

int main() {
    Solution s;
    cout << s.mySqrt(16) << endl;
}